direct.go

```diff
@@ -108,79 +113,138 @@ func (d *direct) loadCandidates(initial map[string][]*Masquerade) {
+	if err != nil {
+		return masqueradesRemain
+	}
+	defer conn.Close()
+
+	if !masqueradeGood(postCheck(conn)) {
+		log.Tracef("Unsuccessful vetting with HEAD request, discarding masquerade")
+		return masqueradesRemain
+	}
+	log.Trace("Finished vetting one")
+	return false
+}
+
+// postCheck does a post with invalid data to verify domain-fronting works
+func postCheck(conn net.Conn) bool {
+	client := &http.Client{
+		Transport: httpTransport(conn, nil),
+	}
+	return doPostCheck(client)
+}
+
+func doPostCheck(client *http.Client) bool {
+	req, _ := http.NewRequest(http.MethodPost, testURL, strings.NewReader("a"))
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Tracef("Unsuccessful vetting with HEAD request, discarding masquerade")
+		return false
+	}
+	resp.Body.Close()
+	if resp.StatusCode != http.StatusAccepted {
+		log.Tracef("Unexpected response status vetting masquerade: %v, %v", resp.StatusCode, resp.Status)
+		return false
+	}
+	return true
+}
+
 // NewDirect creates a new http.RoundTripper that does direct domain fronting.
 func NewDirect(timeout time.Duration) http.RoundTripper {
 	instance, ok := _instance.Get(timeout)
 	if !ok {
 		panic(fmt.Errorf("No DirectHttpClient available within %v", timeout))
 	}
-	return instance.(*direct).NewDirect()
-}
-
-// NewDirect creates a new http.RoundTripper that does direct domain fronting.
-func (d *direct) NewDirect() http.RoundTripper {
-	return &directTransport{
-		Transport: http.Transport{
-			Dial:                d.Dial,
-			TLSHandshakeTimeout: 40 * time.Second,
-			DisableKeepAlives:   true,
-			TLSClientConfig: &tls.Config{
-				ClientSessionCache: clientSessionCache,
-			},
-		},
-	}
+	return instance.(http.RoundTripper)
 }
 
 // Do continually retries a given request until it succeeds because some
 // fronting providers will return a 403 for some domains.
-func (d *direct) Do(req *http.Request) (*http.Response, error) {
-	for i := 0; i < 6; i++ {
-		if resp, err := d.NewDirect().RoundTrip(req); err != nil {
+func (d *direct) RoundTrip(req *http.Request) (*http.Response, error) {
+	var body []byte
+	var err error
+	if req.Body != nil {
+		// store body in-memory to be able to replay it if necessary
+		body, err = ioutil.ReadAll(req.Body)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to read request body: %v", err)
+		}
+	}
+	for i := 0; i < maxTries; i++ {
+		if body != nil {
+			req.Body = ioutil.NopCloser(bytes.NewReader(body))
+		}
+		conn, masqueradeGood, err := d.dial()
+		if err != nil {
+			// unable to find good masquerade, fail
+			return nil, err
+		}
+		tr := httpTransport(conn, clientSessionCache)
+		resp, err := tr.RoundTrip(req)
+		if err != nil {
 			log.Errorf("Could not complete request %v", err)
-		} else if resp.StatusCode > 199 && resp.StatusCode < 400 {
-			return resp, err
-		} else {
-			_ = resp.Body.Close()
+			masqueradeGood(false)
+			continue
 		}
+		resp.Body.Close()
```

Comment:
The body is prematurely closed, which causes `flashlight/config` to fail test. PR soon.